BracesFixer - Fix indentation issues with comments

julienfalque opened the pull request:
Fixes #797, #1196, #1371 and #2539.
Fixes #2176 at least partially as some comments don't provide input code so I couldn't reproduce them.
May fix #1344, to confirm.
Replaces #2611.

SpacePossum commented:
can you add the utests from 2611 (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2611/files) ?

julienfalque commented:
Tests added. Note that I added them to a different provider method as I don't think they were related to `testPreserveLineAfterControlBrace`. All passed without further changes.

I also added test cases (and a fix) for #2608 which is now fixed by this PR as well. Without this last change, this one was only fixed partially: everything was correcly indented, except multiline content of PHPDoc comments.

GrahamCampbell commented:
Looks good as far as I can see. 👍

julienfalque commented:
@GrahamCampbell you reported #797, would it be possible for you to somehow check whether this PR fixes your issue?

GrahamCampbell commented:
My mac is currently broken, so I am unable to run code easily for now. Taking it in for repair today, so will check on this soon. :)

GrahamCampbell commented:
Got myself a working Ubuntu environment in the meanwhile. I can confirm that this PR behaves correctly on the issue that I reported. :+1:

julienfalque commented:
Thanks for checking @GrahamCampbell!

SpacePossum commented:
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
                '<?php
class A
{
    public function B()
    {
        /*
                */
      $a = 1;
    }
}
',
                '<?php
class A {
    public function B()
    {/*
        */
      $a = 1;
    }
}
',
```

Adding the case shows a problem; after fixing and refixing the code changes.

SpacePossum commented:
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
                '<?php
class B
{
    public function B()
    {
        /*
            *//**/
        $a = 1;
    }
}
',
                '<?php
class B {
    public function B()
    {
    /*
        *//**/
       $a = 1;
    }
}
',
            ),
            array(
                '<?php
class C
{
    public function C()
    {
        /* */#
        $a = 1;
    }
}
',
                '<?php
class C {
    public function C()
    {
    /* */#
       $a = 1;
    }
}
',
            ),
```

two more cases, not sure if you want to pick up these as well

SpacePossum commented on src/Fixer/Basic/BracesFixer.php:
```diff
@@ -429,7 +439,11 @@ function ($item) {
                     || !($nextToken->isWhitespace() && $nextToken->isWhitespace(" \t"))
```
while you're at it ;) can you please wrap the `($nextToken [...] && [...] "\n") // preserve blank lines` 
within `(` `)`

and update `// set indent only if it is not a case, when comment is following { in same line` to `on the same line`

julienfalque replied:
Updated the comment.

I added braces while fixing the first test case, is it what you wanted to do?

SpacePossum replied:
yeah, thanks man :)

SpacePossum replied:
last changes undid the request;
I like when a condition in the form;
```php

if ($a || $b && $c) {
    echo 1;
}
```

is written out explicit as;
```php

if ($a || ($b && $c)) {
    echo 1;
}
```

maybe one day I make a fixer for it ;)

SpacePossum replied:
```php
                if (
                    !$nextNonWhitespaceToken->isComment()
                    || (
                        !($nextToken->isWhitespace() && $nextToken->isWhitespace(" \t"))
                        && ($nextToken->isWhitespace() && 1 === substr_count($nextToken->getContent(), "\n")) // preserve blank lines
                    )
                ) {
```

this test still confuses me haha

julienfalque replied:
Actually the condition could be simplified, it should be easier to read now :)

SpacePossum commented:
if you've time, please take a look at the code coverage by the tests as well

julienfalque commented:
Test cases added and fixed.

mooror commented:
I'm currently using the latest PHP-CS-Fixer via its atom package and experiencing comment indenting problems myself. I am excited to see a possible fix well under way. 

Keep up the good work,
Mooror

SpacePossum commented on src/Fixer/Basic/BracesFixer.php:
```diff
@@ -354,13 +354,35 @@ function ($item) {
+                    }
+                }
+
+                if ($nextLineCanBeIndented) {
```
this check can be moved at the end inside the block `if (1 === $nextlevel)` (i.e. remove line `}` at ln 374 and close later)

julienfalque replied:
Done.

SpacePossum commented:
please add;
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
'<?php
if ($a){/*
*/
echo 1;
}
',
```

note, this one is fixed correctly

```php
<?php
if ($a) { /*
*/
echo 1;
}
```

i.e. issue is `{` followed by multi line comment directly

julienfalque commented:
Rebased onto 2.2 and added the test case.

SpacePossum commented:
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
'<?php
if ($a) { /**/ /*
*/
    echo 1;
    echo 2;
}
',
                '<?php
if ($a){ /**/ /*
*/
echo 1;
echo 2;
}
',
```

sorry, gave you the wrong case :(

julienfalque commented:
Tests updated and fixed.

SpacePossum commented on src/Fixer/Basic/BracesFixer.php:
```diff
@@ -354,13 +354,47 @@ function ($item) {
+                            $nextLineCanBeIndented = true;
+                        }
+
+                        if ($nextLineCanBeIndented) {
```
can be merged with 375; `if ($i === $startBraceIndex || nextLineCanBeIndented)`

julienfalque replied:
Done.

SpacePossum replied:
:+1:

SpacePossum commented:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2849#discussion_r125471150
and please throw in some comments to make it easier to follow the code ;)

julienfalque commented:
Fixed the first example you provided. I'm not sure about the second one though: how can you tell the indentation of the second comment line is intended and should not be changed?

SpacePossum commented:
the second comment line is not of concern of the fixer is it? i.e. it is not related to any braces?

julienfalque commented:
It is: the fixer makes sure the braces are properly placed and that their content is properly indented. Here is the fixer's description:
> The body of each structure MUST be enclosed by braces. Braces should be properly placed. Body of braces should be properly indented.

In the example above, the second comment was not moved because of the various _"previous token is not a comment"_ checks that prevented anything that follows a comment to be properly indented, which is what this PR tries to fix.

keradus commented:
the second case I provided shall not be modified (and it's not modified on `2.2@stable`)

julienfalque commented:
> it's not modified on `2.2@stable`

Like I said, anything that follows the first comment is not modified on `2.2@stable`. If you try this:
```php
function foo()
{
    $bar = 1;                   // multiline ...
                                $foo = 3;
    $baz  = 2;                  // next comment
}
```
you'll notice that the `$foo` line isn't modified. With the changes in this PR, it will be, but I see no reason to _not_ fix this line when it is a comment.

SpacePossum commented:
what about skipping any line that is line break + white space* + comment?

julienfalque commented:
Then following case would not be changed:
```php
function foo()
{
        // foo
    return $foo;
}
```
which looks wrong to me.

keradus commented:
the thing is that a lot of users align "single-line" comments to make them be kind-of multiline. breaking it is no-go

keradus commented:
Extra work made here, glory to you @julienfalque !

julienfalque commented:
Thank you guys for the help you provided ;)

SpacePossum commented:
awesome to see this fixed, one of the oldest long standing hard to fix bugs is finally gone :D
thanks @julienfalque !

mooror commented:
Job well done!

MarekLew commented:
don`t work corectly for one line comment


```
abstract class XAbcObject
{
    protected $dbc;

    protected $fields; // aa

  protected $emptyFieldsMessages; // bbb
  protected $pk; // cc
  protected $table; // dd
  protected $values = array(); // ee
  protected $errorMsg;  
    protected $errorField;
....
```

julienfalque commented:
@MarekLew Thanks for your report, can you please open a dedicated issue?
